src/main/java/net/wurstclient/hacks/PortalEspHack.java

```diff
@@ -0,0 +1,281 @@
+		new HashMap<>();
+	private final Set<ChunkPos> chunksToUpdate =
+		Collections.synchronizedSet(new HashSet<>());
+	private ExecutorService threadPool;
+	
+	private boolean groupsUpToDate;
+	
+	public PortalEspHack()
+	{
+		super("PortalESP");
+		setCategory(Category.RENDER);
+		
+		addSetting(style);
+		groups.stream().flatMap(PortalEspBlockGroup::getSettings)
+			.forEach(this::addSetting);
+		addSetting(area);
+	}
+	
+	@Override
+	public void onEnable()
+	{
+		threadPool = MinPriorityThreadFactory.newFixedThreadPool();
+		groupsUpToDate = false;
+		
+		EVENTS.add(UpdateListener.class, this);
+		EVENTS.add(PacketInputListener.class, this);
+		EVENTS.add(CameraTransformViewBobbingListener.class, this);
+		EVENTS.add(RenderListener.class, this);
+		
+		PortalEspRenderer.prepareBuffers();
+	}
+	
+	@Override
+	public void onDisable()
+	{
+		EVENTS.remove(UpdateListener.class, this);
+		EVENTS.remove(PacketInputListener.class, this);
+		EVENTS.remove(CameraTransformViewBobbingListener.class, this);
+		EVENTS.remove(RenderListener.class, this);
+		
+		threadPool.shutdownNow();
+		chunksToUpdate.clear();
+		groups.forEach(PortalEspBlockGroup::clear);
+		PortalEspRenderer.closeBuffers();
+	}
+	
+	@Override
+	public void onReceivedPacket(PacketInputEvent event)
+	{
+		ChunkPos chunkPos = ChunkUtils.getAffectedChunk(event.getPacket());
+		
+		if(chunkPos != null)
+			chunksToUpdate.add(chunkPos);
+	}
+	
+	@Override
+	public void onCameraTransformViewBobbing(
+		CameraTransformViewBobbingEvent event)
+	{
+		if(style.getSelected().hasLines())
+			event.cancel();
+	}
+	
+	@Override
+	public void onUpdate()
+	{
+		ArrayList<Block> blockList =
+			groups.stream().map(PortalEspBlockGroup::getBlock)
+				.collect(Collectors.toCollection(ArrayList::new));
+		
+		DimensionType dimension = MC.world.getDimension();
+		HashSet<ChunkPos> chunkUpdates = clearChunksToUpdate();
+		
+		// remove outdated ChunkSearchers
+		for(ChunkSearcherMulti searcher : new ArrayList<>(searchers.values()))
+		{
+			boolean remove = false;
+			ChunkPos searcherPos = searcher.getPos();
+			
+			// wrong dimension
+			if(dimension != searcher.getDimension())
+				remove = true;
+			
+			// out of range
+			else if(!area.isInRange(searcherPos))
+				remove = true;
+			
+			// chunk update
+			else if(chunkUpdates.contains(searcherPos))
+				remove = true;
+			
+			if(remove)
+			{
+				groupsUpToDate = false;
+				searchers.remove(searcherPos);
+				searcher.cancelSearching();
+			}
+		}
+		
+		// add new ChunkSearchers
+		for(Chunk chunk : area.getChunksInRange())
+		{
+			ChunkPos chunkPos = chunk.getPos();
+			if(searchers.containsKey(chunkPos))
+				continue;
+			
+			groupsUpToDate = false;
+			ChunkSearcherMulti searcher =
+				new ChunkSearcherMulti(chunk, blockList, dimension);
+			searchers.put(chunkPos, searcher);
+			searcher.startSearching(threadPool);
+		}
+		
+		if(!areAllChunkSearchersDone())
+			return;
+		
+		if(!groupsUpToDate)
+			updateGroupBoxes();
```

Comment:
Instead of breaking this code down into smaller methods within `PortalEspHack`, my plan is to move it into a separate `ChunkSearcherCoordinator` class. This is because other hacks like Search and CaveFinder have almost identical code in their `onUpdate()` methods, which could all be combined into a central place.

However, there are a few differences between these hacks that need to be considered when making a `ChunkSearcherCoordinator`:

First, Search's `onUpdate()` method has an additional check that is not present in other hacks:
```java
// wrong block
if(currentBlock != searcher.getBlock())
	remove = true;
```

Second, PortalESP uses `ChunkSearcherMulti` instead of `ChunkSearcher`.

The second difference could be addressed by modifying the `ChunkSearcher` class so that it can always handle multiple different block types, and deleting the `ChunkSearcherMulti` class. However, depending on how this is implemented, it might complicate the first difference because the updated `ChunkSearcher` would no longer have a single target block to compare against in Search.

If the updated `ChunkSearcher` class were designed with an `ArrayList<Block>` of target blocks, as is currently the case with `ChunkSearcherMulti`, then this list would be easy enough to compare with Search's `currentBlock` variable - just get the only entry in the list and compare that.

But I do think there is value in making `ChunkSearcher` more flexible by designing it with a `BiPredicate<BlockPos, BlockState>` instead. This BiPredicate would act as a query of what to look for in the blocks and could allow searching for any arbitrary properties or patterns.

But then, the `ChunkSearcher`s would no longer have a target block to compare against in Search. And since `BiPredicate`s are not necessarily comparable with each other, creating a new query from `currentBlock` and comparing that against the `ChunkSearcher`'s existing query might not work either.

...and this is what I'm currently pondering. Let me know if you have any suggestions.

Reply:
Done. @nobotreply